hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientPartitionServiceImpl.java

```diff
@@ -179,33 +169,68 @@ private boolean isClusterFormedByOnlyLiteMembers(Collection<Member> memberList)
                                           int partitionStateVersion,
                                           boolean partitionStateVersionExist) {
-        if (partitions.isEmpty()) {
-            if (logger.isFinestEnabled()) {
-                logger.finest("Partition response is empty, state version:"
-                        + (partitionStateVersionExist ? partitionStateVersion : "NotAvailable"));
-            }
-            return;
-        }
 
-        synchronized (lock) {
-            if (!partitionStateVersionExist || partitionStateVersion > lastPartitionStateVersion) {
-                Map<Integer, Address> partitionToAddressMap = convertToPartitionToAddressMap(partitions);
-                this.partitions = Collections.unmodifiableMap(partitionToAddressMap);
+        while (true) {
+            PartitionTable current = this.partitionTable.get();
+            if (!shouldBeApplied(connection, partitions, partitionStateVersion, partitionStateVersionExist, current)) {
+                return;
+            }
+            Map<Integer, Address> newPartitions = Collections.unmodifiableMap(convertToPartitionToAddressMap(partitions));
+            PartitionTable newMetaData = new PartitionTable(connection, partitionStateVersion, newPartitions);
+            if (this.partitionTable.compareAndSet(current, newMetaData)) {
+                // partition count is set once at the start. Even if we reset the partition table when switching cluster
+                //we want to remember the partition count. That is why it is a different field.
                 if (partitionCount == 0) {
-                    partitionCount = this.partitions.size();
+                    partitionCount = newPartitions.size();
                 }
-                lastPartitionStateVersion = partitionStateVersion;
                 if (logger.isFinestEnabled()) {
                     logger.finest("Processed partition response. partitionStateVersion : "
                             + (partitionStateVersionExist ? partitionStateVersion : "NotAvailable")
-                            + ", partitionCount :" + this.partitions.size());
+                            + ", partitionCount :" + newPartitions.size() + ", connection : " + connection);
                 }
+                return;
+            }
 
+        }
+    }
+
+    private boolean shouldBeApplied(Connection connection, Collection<Map.Entry<Address, List<Integer>>> partitions,
+                                    int partitionStateVersion, boolean partitionStateVersionExist, PartitionTable current) {
+        if (partitions.isEmpty()) {
+            if (logger.isFinestEnabled()) {
+                logFailure(connection, partitionStateVersion, partitionStateVersionExist, current,
+                        "response is empty");
+            }
+            return false;
+        }
+        if (!current.connection.equals(connection)) {
```

Comment:
What if `current.connection` is null? is that possible?

Reply:
Hmm. I should do the check other way around. The connection on the right hand side can not be null